app/models/course_template.rb

```diff
@@ -0,0 +1,122 @@
+# Course stub to be copied by teachers for their own organisations
+
+require 'net/http'
+require 'course_refresher'
+
+class CourseTemplate < ActiveRecord::Base
+  include SystemCommands
+
+  validates :name,
+            presence: true,
+            uniqueness: true,
+            length: { within: 1..40 },
+            format: {
+              without: / /,
+              message: 'should not contain white spaces'
+            }
+  validates :title,
+            presence: true,
+            length: { within: 1..40 }
+  validates :description, length: { maximum: 512 }
+  validates :git_branch, presence: true
+  validates :source_url, presence: true
+  validate :valid_git_repo?
+  validate :valid_source_backend?
+
+  after_initialize :set_default_source_backend
+
+  has_many :courses
+
+  scope :not_expired, -> { where('expires_at IS NULL OR expires_at > ?', Time.now) }
+  scope :not_hidden, -> { where(hidden: false) }
+  scope :hidden, -> { where(hidden: true) }
+  scope :not_dummy, -> { where(dummy: false) }
+  scope :available, -> { not_expired.not_hidden.not_dummy }
+
+  after_save :update_courses_sourcedata
+
+  after_destroy :delete_courses
+  after_destroy :delete_cache
+
+  def self.new_dummy(course, options = {})
+    CourseTemplate.new(dummy: true,
+                       name: course.name,
+                       title: course.title,
+                       description: course.description,
+                       material_url: course.material_url,
+                       source_url: options[:source_url],
+                       git_branch: options[:git_branch] || 'master',
+                       source_backend: options[:source_backend] || 'git')
+  end
+
+  def valid_git_repo?
+    return true unless source_url_changed? || git_branch_changed? # don't attempt repo cloning if source url or branch wasn't even changed
+    Dir.mktmpdir do |dir|
+      sh!('git', 'clone', '-q', '-b', git_branch, source_url, dir)
+      File.exist?("#{dir}/.git")
+    end
```

Comment:
We can btw make this faster with git ls-remote

` git ls-remote git@github.com:testmycode/tmc-server.git`
and then just see whether exit code is 0.